File: klippy/extras/crash_detect.py
# StallGuard crash detection (host side).  Pairs with src/crash_detect.c.
#
# Prusa Core One firmware ships "Crash Detection" (default-on): the TMC2130 X/Y
# drivers flag a motor stall via StallGuard during moves and the printer halts so
# the nozzle stops grinding into the bed/part.  Our Klipper port had NO such guard
# on a plain G1 move (the loadcell only protects homing/probing).  This module is
# the equivalent: while ARMED (for the duration of a print) the MCU reads sg_result
# over SPI every sample_period and, when it stays at/near 0 (Prusa's DIAG1 comparator
# fires at SG_RESULT==0), calls shutdown() -- an immediate halt of all steppers + heaters.
#
# DETECTION + STOP only: Prusa's Crash_s also does re-home + g-code replay recovery,
# which Klipper has no infrastructure for.  After a crash: FIRMWARE_RESTART + re-print.
#
# Trigger = an ABSOLUTE sg_result <= sg_floor (==0 with SGT biasing where 0 lands),
# matching Prusa's CRASH_STALL_GUARD (SGT +2) + CRASH_FILTER (sfilt off), NOT the
# earlier % -drop-from-baseline (which false-tripped: sg_result is low just above the
# gate and climbs with speed, so a % drop reads fast moves as stalls).  A VELOCITY GATE
# (TSTEP) still skips samples below the coolStep velocity, where sg_result is invalid-
# low, so accel/decel/slow moves can't false-trip.  The gate value is taken from the
# driver's own configured TCOOLTHRS (coolstep_threshold) -- no clock math.
#
#   [crash_detect stepper_x]
#     #tmc: tmc2130 stepper_x   # default; the driver to read sg_result from
#     sg_floor: 2               # stall when sg_result stays <= this (Prusa SG==0) ...
#     sample_count: 4           # ... for this many consecutive samples ...
#     sample_period: 0.0003     # ... spaced this far apart (s)
#     crash_sgt: 2              # TMC SGT applied on arm (Prusa +2; higher = less sensitive)
#
#   CRASH_DETECT ENABLE=1   ; arm  (PRINT_START, after homing+mesh+prime)
#   CRASH_DETECT ENABLE=0   ; disarm (PRINT_END / CANCEL / abort paths)
#
# NOTE: place [crash_detect stepper_x] AFTER [tmc2130 stepper_x] in the config so the
# driver object exists when this binds to it.
import collections


class CrashDetect:
    def __init__(self, config):
        self.printer = config.get_printer()
        self.name = config.get_name().split()[-1]          # e.g. "stepper_x"
        self.tmc_name = config.get('tmc', 'tmc2130 ' + self.name)
        # Prusa SG==0 trigger: stall when sg_result stays AT/below this absolute floor.
        # With SGT biasing where 0 lands, near-0 = a real stall at any speed (velocity-
        # robust), unlike a % drop from a baseline. 0 = exact Prusa DIAG1 (SG_RESULT==0);
        # a small floor adds margin for the SPI-polled read.
        self.sg_floor = config.getint('sg_floor', 2, minval=0, maxval=1023)
        # consecutive at-floor samples required -- debounces a stray single-sample 0
        # (DIAG/line noise). Prusa uses a single hardware DIAG1 edge; a few SPI samples
        # is the polled equivalent. sample_count*sample_period = the sustained window.
        self.sample_count = config.getint('sample_count', 4,
                                          minval=1, maxval=255)
        self.sample_period = config.getfloat('sample_period', 0.0003, above=0.)
        # CRASH_STALL_GUARD: the TMC2130 SGT applied on arm (separate from homing SGT).
        # Higher = LESS sensitive. Prusa = +2 on X/Y. Restored on disarm.
        self.crash_sgt = config.getint('crash_sgt', 2, minval=-64, maxval=63)
        self.mcu = self.spi = self.mcu_tmc = self.fields = self.tmc = None
        self._cmd_queue = self._arm_cmd = self._query_cmd = None
        self._armed = False
        self._prev = collections.OrderedDict()
        self._dirty = collections.OrderedDict()
        self._bind_tmc()
        self._oid = self.mcu.create_oid()
        self.mcu.register_config_callback(self._build_config)
        # One shared CRASH_DETECT command drives every [crash_detect ...] instance;
        # the first-created instance registers it, the rest piggy-back.
        gcode = self.printer.lookup_object('gcode')
        try:
            gcode.register_command('CRASH_DETECT', self._cmd_CRASH_DETECT,
                                   desc=self.cmd_CRASH_DETECT_help)
        except self.printer.config_error:
            pass        # already registered by a sibling instance

    def _bind_tmc(self):
        if self.mcu is not None:
            return
        tmc = self.printer.lookup_object(self.tmc_name)
        self.tmc = tmc
        self.mcu_tmc = tmc.mcu_tmc
        self.fields = self.mcu_tmc.get_fields()
        self.spi = self.mcu_tmc.tmc_spi.spi
        self.mcu = self.spi.get_mcu()

    def _build_config(self):
        spi_oid = self.spi.get_oid()
        self.mcu.add_config_cmd("config_crash_detect oid=%d spi_oid=%d"
                                % (self._oid, spi_oid))
        # disarm on (re)start so a FIRMWARE_RESTART after a crash comes up unarmed
        self.mcu.add_config_cmd(
            "crash_detect_arm oid=%d clock=0 rest_ticks=0 sample_count=0"
            " sg_floor=0 gate_tstep=0" % (self._oid,),
            on_restart=True)
        self._cmd_queue = self.mcu.alloc_command_queue()
        self._arm_cmd = self.mcu.lookup_command(
            "crash_detect_arm oid=%c clock=%u rest_ticks=%u sample_count=%c"
            " sg_floor=%hu gate_tstep=%u",
            cq=self._cmd_queue)
        self._query_cmd = self.mcu.lookup_query_command(
            "crash_detect_query_state oid=%c",
            "crash_detect_state oid=%c armed=%c",
            oid=self._oid, cq=self._cmd_queue)

    # --- TMC register save/restore (mirrors sg_endstop / TMCVirtualPinHelper) ---
    def _set_field(self, name, value):
        if name not in self._prev:
            self._prev[name] = self.fields.get_field(name)
        reg = self.fields.lookup_register(name)
        self._dirty[reg] = self.fields.set_field(name, value)

    def _send(self):
        for reg, val in self._dirty.items():
            self.mcu_tmc.set_register(reg, val)
        self._dirty.clear()

    def _set_tmc_check(self, enable):
        # Suspend Klipper's periodic TMC error-check (the ~1 Hz GSTAT poll) on THIS
        # driver while crash detection is armed. The MCU's own sg_result reads use the
        # TMC2130's PIPELINED SPI read latch (a read returns the register addressed in
        # the PREVIOUS transfer); a concurrent host GSTAT poll then reads back the wrong
        # register and Klipper sees a bogus "reset=1" -> driver-error shutdown. (Same
        # shared-spi3a collision the phase-stepping executor hit.) Also no-op
        # get/set_register so a stray host access (e.g. idle-timeout disable) can't
        # collide. All restored on disarm. enable=True restores, False suspends.
        try:
            cmdhelper = self.tmc.get_status.__self__
            ec = getattr(cmdhelper, 'echeck_helper', cmdhelper)
            mt = self.mcu_tmc
            if enable:
                if hasattr(ec, '_cd_saved_start'):
                    ec.start_checks = ec._cd_saved_start
                    del ec._cd_saved_start
                if hasattr(mt, '_cd_saved_get'):
                    mt.get_register = mt._cd_saved_get
                    mt.set_register = mt._cd_saved_set
                    del mt._cd_saved_get, mt._cd_saved_set
                ec.start_checks()
            else:
                ec.stop_checks()
                # idempotence: if suspend ever runs twice, don't re-save the no-op shadow
                # (that would lose the REAL get/set_register until klippy restart). self._armed
                # guards double-arm today; this is defense in depth.
                if not hasattr(ec, '_cd_saved_start'):
                    ec._cd_saved_start = ec.start_checks
                    ec.start_checks = lambda *a, **k: False
                if not hasattr(mt, '_cd_saved_get'):
                    mt._cd_saved_get = mt.get_register
                    mt._cd_saved_set = mt.set_register
                    # benign read value: DRV_STATUS cs_actual=0x1F, GSTAT reset/drv_err/uv=0
                    mt.get_register = lambda *a, **k: 0x001f0000
                    mt.set_register = lambda *a, **k: None
        except Exception:
            pass

    def arm(self):
        if self._armed:
            return
        # Match Prusa's crash StallGuard config with REAL TMC writes BEFORE suspending the
        # checks (which no-op set_register); all restored from self._prev on disarm:
        #  - en_pwm_mode=0  : spreadCycle (stealthChop has no StallGuard)
        #  - sgt=crash_sgt  : CRASH_STALL_GUARD (+2), less sensitive than homing
        #  - sfilt=0        : CRASH_FILTER off (per-fullstep, no 4-sample averaging)
        self._set_field("en_pwm_mode", 0)
        self._set_field("sgt", self.crash_sgt)
        self._set_field("sfilt", 0)
        self._send()
        # gate = the driver's own coolStep/StallGuard TSTEP window (coolstep_threshold),
        # read from the cached field (no SPI): the MCU skips a sample when TSTEP > gate.
        gate = self.fields.get_field("tcoolthrs")
        toolhead = self.printer.lookup_object('toolhead')
        print_time = toolhead.get_last_move_time()
        clock = self.mcu.print_time_to_clock(print_time)
        rest_ticks = self.mcu.seconds_to_clock(self.sample_period)
        # suspend host TMC polling on this driver BEFORE the MCU starts reading sg_result
        self._set_tmc_check(False)
        self._arm_cmd.send(
            [self._oid, clock, rest_ticks, self.sample_count, self.sg_floor, gate],
            reqclock=clock)
        self._armed = True

    def disarm(self):
        if not self._armed:
            return
        # stop the MCU's SPI reads first, then confirm (a query round-trip) the disarm
        # was processed before resuming host polling -- so the two never overlap on spi3a.
        self._arm_cmd.send([self._oid, 0, 0, 0, 0, 0])
        if not self.mcu.is_fileoutput():
            self._query_cmd.send([self._oid])
        self._set_tmc_check(True)           # restore host TMC polling (get/set real again)
        for field, val in list(self._prev.items()):   # restore en_pwm_mode (real write now)
            self._set_field(field, val)
        self._send()
        self._prev.clear()
        self._armed = False

    cmd_CRASH_DETECT_help = ("Arm/disarm StallGuard crash detection on X/Y for the"
                             " duration of a print (CRASH_DETECT ENABLE=1 | ENABLE=0)")

    def _cmd_CRASH_DETECT(self, gcmd):
        # Klipper extended gcode needs KEY=VALUE -- a bare "ON"/"OFF" token is a
        # malformed command, so arm/disarm via ENABLE=1/0 (default ENABLE=1).
        enable = gcmd.get_int('ENABLE', 1, minval=0, maxval=1)
        # (shared-SPI3 hazard): crash_detect's task-level TMC sg_result reads run on the SAME
        # SPI3 the phase-stepping executor owns. While phase-stepping is engaged (direct_mode) a
        # blocking read can collide mid-frame with the prio-0 TIM8/DMA XDIRECT write -> a corrupted
        # coil vector; and it is INERT anyway (no step pulses -> TSTEP pegs -> the velocity gate
        # never passes). So refuse to ARM while any phase_exec is engaged: lose nothing, remove the
        # hazard. (Disarm is always allowed; disarm() is a no-op if not armed.) PRINT_START engages
        # phase-stepping BEFORE arming crash detect, so this guard sees it.
        if enable and any(getattr(o, '_trapq_engaged', False)
                          for _n, o in self.printer.lookup_objects('phase_exec')):
            gcmd.respond_info("Crash detection NOT armed: phase-stepping engaged "
                              "(shares SPI3; StallGuard gate inert in direct_mode)")
            return
        for _name, obj in self.printer.lookup_objects('crash_detect'):
            if enable:
                obj.arm()
            else:
                obj.disarm()
        gcmd.respond_info("Crash detection %s" % ("ON" if enable else "OFF"))


def load_config_prefix(config):
    return CrashDetect(config)

File: src/stm32/crash_detect.c
// StallGuard crash detection (TMC2130) — the Prusa "Crash Detection" equivalent.
//
// A plain G1 print move has NO crash protection in our port (the loadcell only
// guards homing/probing). Prusa's Buddy fw flags a motor stall (= a nozzle/gantry
// crash) via StallGuard during moves and halts. This module does the same: while
// ARMED (for the duration of a print), it reads the TMC2130 sg_result over SPI and,
// when it stays at/near 0 (an absolute stall, the way Prusa's DIAG1 comparator fires
// at SG_RESULT==0), calls shutdown() — an immediate ISR-speed halt of all steppers/
// heaters. The nozzle stops grinding.
//
// This is DETECTION + STOP only (no Prusa-style re-home + g-code replay recovery:
// Klipper has no replay infra). After a crash: FIRMWARE_RESTART + re-print.
//
// Detection matches Prusa's production design (Prusa-Firmware-Buddy CRASH_STALL_GUARD
// / CRASH_FILTER): trigger on an ABSOLUTE sg_result <= sg_floor (==0, with the SGT
// register biasing where 0 lands), NOT a relative fall from a baseline. The earlier
// adaptive-baseline/% -drop design false-tripped because sg_result is intrinsically
// low just above the velocity gate and climbs with speed, so a % drop reads fast moves
// as stalls; an absolute near-0 is velocity-robust. Two notes:
//  1) the trigger is shutdown() (whole-print safety) instead of a homing trsync;
//  2) a VELOCITY GATE: StallGuard's sg_result is only valid above the coolStep
//     velocity (TSTEP <= TCOOLTHRS). Below it (slow moves, accel/decel), sg_result
//     is invalid-low; we skip those samples so they can't false-trigger a crash.
//     (Same velocity limitation Prusa's crash detection has.)
// SGT=+2 and sfilt=off are applied by the host (extras/crash_detect.py) on arm.
//
// The SPI read runs in a task (woken by the timer) because spidev_transfer is not
// IRQ-safe.

#include "basecmd.h"   // oid_alloc, oid_lookup, oid_next
#include "board/irq.h" // irq_disable
#include "board/misc.h" // timer_read_time
#include "command.h"   // DECL_COMMAND, sendf, shutdown
#include "sched.h"     // struct timer, sched_add_timer, DECL_TASK
#include "spicmds.h"   // spidev_transfer, spidev_oid_lookup

#define TMC_TSTEP       0x12   // read address: TSTEP (velocity, 20-bit; large = slow)
#define TMC_DRV_STATUS  0x6f   // read address: DRV_STATUS (sg_result in bits 9:0)

struct crash_detect {
    struct timer time;
    uint32_t rest_time;
    uint32_t gate_tstep;      // velocity gate: skip a sample when TSTEP > this (slower
                              // than the coolStep gate). 0 = gate disabled.
    struct spidev_s *spi;
    uint16_t sg_floor;        // Prusa SG==0 trigger: stall when sg_result <= this
                              // absolute floor (SGT biases where 0 lands). NOT a % drop.
    uint8_t flags, sample_count, trigger_count;
    // --- SHADOW telemetry: when armed with shadow=1, accumulate the gated-in
    // sg_result distribution instead of shutting down, so sg_floor / crash_sgt can
    // be tuned from real prints (the analog of estall's I2 shadow histogram, but
    // MCU-side because sg_result is read here, not streamed to the host). Reset on a
    // shadow arm, reported to the host on disarm.
    uint16_t s_min;           // min gated-in sg_result (0xffff = no gated samples)
    uint32_t s_samples;       // gated-in sample count
    uint32_t s_le[7];         // cumulative count sg_result <= {0,2,5,10,20,50,100}
    uint32_t s_wtrip;         // times the sustained-stall trigger WOULD have fired
};

enum { CDF_ARMED = 1 << 0, CDF_SHADOW = 1 << 1 };

static const uint16_t cd_le_floor[7] = { 0, 2, 5, 10, 20, 50, 100 };

static struct task_wake cd_wake;

// forward decl so the task can use it as the oid key
void command_config_crash_detect(uint32_t *args);

// Read TSTEP and DRV_STATUS(sg_result) from the TMC2130. TMC SPI reads are
// "delayed": a transfer returns the value of the register addressed in the
// PREVIOUS transfer. Three self-priming transfers therefore yield both fresh:
//   T1 addr=TSTEP       (returns stale -- ignored)
//   T2 addr=DRV_STATUS  (returns TSTEP, addressed in T1)
//   T3 addr=DRV_STATUS  (returns DRV_STATUS, addressed in T2)
static void
cd_read(struct crash_detect *e, uint32_t *tstep, uint16_t *sg)
{
    uint8_t buf[5];
    buf[0] = TMC_TSTEP; buf[1] = buf[2] = buf[3] = buf[4] = 0;
    spidev_transfer(e->spi, 1, sizeof(buf), buf);          // addr TSTEP (return stale)
    buf[0] = TMC_DRV_STATUS; buf[1] = buf[2] = buf[3] = buf[4] = 0;
    spidev_transfer(e->spi, 1, sizeof(buf), buf);          // return = TSTEP
    *tstep = (((uint32_t)buf[1] << 24) | ((uint32_t)buf[2] << 16)
              | ((uint32_t)buf[3] << 8) | buf[4]) & 0xfffff;
    buf[0] = TMC_DRV_STATUS; buf[1] = buf[2] = buf[3] = buf[4] = 0;
    spidev_transfer(e->spi, 1, sizeof(buf), buf);          // return = DRV_STATUS
    *sg = (((uint16_t)buf[3] & 0x03) << 8) | buf[4];        // sg_result = bits 9:0
}

// Timer: wake the task at the sample rate (the SPI read must run in task context).
static uint_fast8_t
crash_detect_event(struct timer *t)
{
    struct crash_detect *e = container_of(t, struct crash_detect, time);
    if (!(e->flags & CDF_ARMED))
        return SF_DONE;
    sched_wake_task(&cd_wake);
    e->time.waketime += e->rest_time;
    return SF_RESCHEDULE;
}

// Task: for every armed detector, read sg_result, apply the velocity gate + the
// adaptive sustained-stall filter, and shutdown() on a sustained stall.
void
crash_detect_task(void)
{
    if (!sched_check_wake(&cd_wake))
        return;
    uint8_t oid = 0xff;
    struct crash_detect *e;
    while ((e = oid_next(&oid, command_config_crash_detect))) {
        if (!(e->flags & CDF_ARMED))
            continue;
        uint32_t tstep;
        uint16_t sg;
        cd_read(e, &tstep, &sg);
        // Velocity gate: sg_result is only valid above the coolStep velocity
        // (TSTEP <= TCOOLTHRS). Below it -- slow moves, accel/decel, and the
        // fast->slow transition that would otherwise read as a sudden drop --
        // skip and reset, so they can't false-trigger and the baseline isn't
        // poisoned with invalid-low values.
        if (e->gate_tstep && tstep > e->gate_tstep) {
            e->trigger_count = e->sample_count;
            continue;
        }
        // SHADOW: accumulate the gated-in distribution (every valid sample), then fall
        // through to the trigger logic (which counts would-trips instead of shutting
        // down while CDF_SHADOW is set).
        if (e->flags & CDF_SHADOW) {
            e->s_samples++;
            if (sg < e->s_min)
                e->s_min = sg;
            for (uint8_t i = 0; i < 7; i++)
                if (sg <= cd_le_floor[i])
                    e->s_le[i]++;
        }
        // Prusa "Crash Detection" trigger: an ABSOLUTE sg_result <= sg_floor (==0 with
        // SGT biasing where 0 lands), NOT a relative fall from a baseline. The TMC2130
        // DIAG1 comparator Prusa uses fires at SG_RESULT==0; reading the same sg_result
        // over SPI and testing <= sg_floor is the software equivalent and is velocity-
        // ROBUST -- unlike a % drop, an absolute near-0 means a real stall at any speed
        // above the gate, so faster moves / higher accel don't false-trip. The small
        // sample_count debounces a stray single-sample 0 (DIAG/line noise).
        if (sg <= e->sg_floor) {
            uint8_t count = e->trigger_count - 1;
            if (!count) {
                if (e->flags & CDF_SHADOW) {
                    e->s_wtrip++;                     // would have tripped -- record, don't halt
                    e->trigger_count = e->sample_count;   // re-arm for the next sustained stall
                    continue;
                }
                shutdown("Crash detected: StallGuard stall");   // __noreturn ISR-speed halt
            }
            e->trigger_count = count;
        } else {
            e->trigger_count = e->sample_count;   // reset on any non-stall
        }
    }
}
DECL_TASK(crash_detect_task);

void
command_config_crash_detect(uint32_t *args)
{
    struct crash_detect *e = oid_alloc(args[0], command_config_crash_detect,
                                       sizeof(*e));
    e->spi = spidev_oid_lookup(args[1]);
}
DECL_COMMAND(command_config_crash_detect, "config_crash_detect oid=%c spi_oid=%c");

// Send the accumulated shadow-telemetry distribution to the host (on disarm of a
// shadow arm). Its format string is auto-collected into the message dict by the build.
static void
crash_detect_report(uint8_t oid, struct crash_detect *e)
{
    sendf("crash_detect_shadow oid=%c samples=%u min=%hu le0=%u le2=%u le5=%u"
          " le10=%u le20=%u le50=%u le100=%u wtrip=%u",
          oid, e->s_samples, e->s_min, e->s_le[0], e->s_le[1], e->s_le[2],
          e->s_le[3], e->s_le[4], e->s_le[5], e->s_le[6], e->s_wtrip);
}

// Arm (sample_count>0) or disarm (sample_count=0) continuous crash detection.
void
command_crash_detect_arm(uint32_t *args)
{
    struct crash_detect *e = oid_lookup(args[0], command_config_crash_detect);
    sched_del_timer(&e->time);
    e->time.waketime = args[1];
    e->rest_time = args[2];
    e->sample_count = args[3];
    if (!e->sample_count) {              // disarm
        if (e->flags & CDF_SHADOW)
            crash_detect_report(args[0], e);   // report the accumulated distribution
        e->flags = 0;
        return;
    }
    e->sg_floor = args[4];
    e->gate_tstep = args[5];
    e->trigger_count = e->sample_count;
    e->flags = CDF_ARMED;
    if (args[6]) {                       // SHADOW arm: accumulate stats, never shutdown
        e->flags |= CDF_SHADOW;
        e->s_min = 0xffff;
        e->s_samples = 0;
        e->s_wtrip = 0;
        for (uint8_t i = 0; i < 7; i++)
            e->s_le[i] = 0;
    }
    e->time.func = crash_detect_event;
    sched_add_timer(&e->time);
}
DECL_COMMAND(command_crash_detect_arm,
             "crash_detect_arm oid=%c clock=%u rest_ticks=%u sample_count=%c"
             " sg_floor=%hu gate_tstep=%u shadow=%c");

void
command_crash_detect_query_state(uint32_t *args)
{
    uint8_t oid = args[0];
    struct crash_detect *e = oid_lookup(oid, command_config_crash_detect);
    sendf("crash_detect_state oid=%c armed=%c", oid, !!(e->flags & CDF_ARMED));
}
DECL_COMMAND(command_crash_detect_query_state, "crash_detect_query_state oid=%c");
